src/CredentialSource/AwsNativeSource.php

```diff
@@ -0,0 +1,360 @@
+     */
+    public static function getRegionFromUrl(callable $httpHandler, string $regionUrl, array $headers): string
+    {
+        // get the region/zone from the region URL
+        $regionRequest = new Request('GET', $regionUrl, $headers);
+        $regionResponse = $httpHandler($regionRequest);
+
+        // Remove last character. For example, if us-east-2b is returned,
+        // the region would be us-east-2.
+        return substr((string) $regionResponse->getBody(), 0, -1);
+    }
+
+    /**
+     * @internal
+     *
+     * @param callable $httpHandler
+     * @param string $securityCredentialsUrl
+     * @param array<string, string|string[]> $headers Request headers to send in with the request.
+     */
+    public static function getRoleName(callable $httpHandler, string $securityCredentialsUrl, array $headers): string
+    {
+        // Get the AWS role name
+        $roleRequest = new Request('GET', $securityCredentialsUrl, $headers);
+        $roleResponse = $httpHandler($roleRequest);
+        $roleName = (string) $roleResponse->getBody();
+
+        return $roleName;
+    }
+
+    /**
+     * @internal
+     *
+     * @param callable $httpHandler
+     * @param string $securityCredentialsUrl
+     * @param array<string, string|string[]> $headers Request headers to send in with the request.
+     * @return array{string, string, ?string}
+     */
+    public static function getSigningVarsFromUrl(
+        callable $httpHandler,
+        string $securityCredentialsUrl,
+        string $roleName,
+        array $headers
+    ): array {
+        // Get the AWS credentials
+        $credsRequest = new Request(
+            'GET',
+            $securityCredentialsUrl . '/' . $roleName,
+            $headers
+        );
+        $credsResponse = $httpHandler($credsRequest);
+        $awsCreds = json_decode((string) $credsResponse->getBody(), true);
+        return [
+            $awsCreds['AccessKeyId'], // accessKeyId
+            $awsCreds['SecretAccessKey'], // secretAccessKey
+            $awsCreds['Token'], // token
+        ];
+    }
+
+    /**
+     * @internal
+     *
+     * @return array{string, string, ?string}
+     */
+    public static function getSigningVarsFromEnv(): ?array
+    {
+        $accessKeyId = getenv('AWS_ACCESS_KEY_ID');
+        $secretAccessKey = getenv('AWS_SECRET_ACCESS_KEY');
+        if ($accessKeyId && $secretAccessKey) {
+            return [
+                $accessKeyId,
+                $secretAccessKey,
+                getenv('AWS_SESSION_TOKEN') ?: null, // session token (can be null)
+            ];
+        }
+
+        return null;
+    }
+
+    /**
+     * Return HMAC hash in binary string
+     */
+    private static function hmacSign(string $key, string $msg): string
+    {
+        return hash_hmac('sha256', self::utf8Encode($msg), $key, true);
+    }
+
+    /**
+     * @TODO add a fallback when mbstring is not available
```

Comment:
Seems like the function `mb_convert_encoding` is present in all supported PHP versions, [ref](https://www.php.net/manual/en/function.mb-convert-encoding.php)

Reply:
See the [installation page](https://www.php.net/manual/en/mbstring.installation.php):

```
mbstring is a non-default extension. This means it is not enabled by default. You must explicitly enable the module with the configure option.
```
It's possible to compile PHP without the MB extension, or to have the extension compiled but not enabled, so we still want to include a fallback.